[Flaky test] TestDefaultHarvesterGroup/assert_a_harvester_for_same_source_cannot_be_started

## Flaky Test

* **Test Name:** TestDefaultHarvesterGroup/assert_a_harvester_for_same_source_cannot_be_started
* **Link:** https://github.com/elastic/beats/blob/285b0d073618e1d34353e9a9bd995ca7dbcf9b7d/filebeat/input/filestream/internal/input-logfile/harvester_test.go#L239
* **Branch:** 9.1
* **Artifact Link:** https://github.com/elastic/beats/actions/runs/16839584778/job/47707354002?pr=45880#step:3:434
* **Notes:** Test times out. I managed to reproduce it on MacOS 15.6 using our `stresstest.sh`

### Stack Trace

```
=== Failed
=== FAIL: filebeat/input/filestream/internal/input-logfile TestDefaultHarvesterGroup (unknown)
--- PASS: TestDefaultHarvesterGroup/assert_a_harvester_is_started_in_a_goroutine (0.00s)
--- PASS: TestDefaultHarvesterGroup/assert_a_harvester_is_only_started_if_harvester_limit_haven't_been_reached (0.00s)
--- PASS: TestDefaultHarvesterGroup/assert_a_harvester_can_be_stopped_and_removed_from_bookkeeper (0.03s)
panic: test timed out after 10m0s
	running tests:
		TestDefaultHarvesterGroup (10m0s)
		TestDefaultHarvesterGroup/assert_a_harvester_for_same_source_cannot_be_started (10m0s)

goroutine 57 [running]:
testing.(*M).startAlarm.func1()
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:2484 +0x308
created by time.goFunc
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/time/sleep.go:215 +0x38

goroutine 1 [chan receive]:
testing.(*T).Run(0x14000183500, {0x10381d7d0?, 0x1400039fb38?}, 0x103b75348)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1859 +0x388
testing.runTests.func1(0x14000183500)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:2279 +0x40
testing.tRunner(0x14000183500, 0x1400039fc68)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1792 +0xe4
testing.runTests(0x14000260dc8, {0x10425a300, 0x18, 0x18}, {0x14000374280?, 0x7?, 0x1042c3480?})
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:2277 +0x3ec
testing.(*M).Run(0x140002ca960)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:2142 +0x588
main.main()
	_testmain.go:91 +0x90

goroutine 50 [chan receive]:
testing.(*T).Run(0x140003e4540, {0x10383af10?, 0x14000080f58?}, 0x1400011c060)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1859 +0x388
github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile.TestDefaultHarvesterGroup(0x140003e4540)
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/input-logfile/harvester_test.go:238 +0x19c
testing.tRunner(0x140003e4540, 0x103b75348)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 1
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1851 +0x374

goroutine 68 [sleep]:
time.Sleep(0x2dc6c0)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/runtime/time.go:338 +0x158
github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile.TestDefaultHarvesterGroup.func6(0x14000102540)
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/input-logfile/harvester_test.go:253 +0x258
testing.tRunner(0x14000102540, 0x1400011c060)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 50
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1851 +0x374

goroutine 69 [chan receive]:
github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile.blockUntilCancelOnRun({0x1400011c1f0, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {{0x0, 0x0}, {0x0, ...}, ...}, ...}, ...)
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/input-logfile/harvester_test.go:443 +0x30
github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile.(*mockHarvester).Run(_, {0x1400011c1f0, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {{0x0, 0x0}, ...}, ...}, ...)
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/input-logfile/harvester_test.go:422 +0x1e4
github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile.startHarvester.func1({0x103b88748, 0x140001320a0})
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/input-logfile/harvester.go:237 +0x7c0
github.com/elastic/beats/v7/filebeat/input/filestream/internal/task.(*Group).Go.func1()
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/task/group.go:114 +0x124
created by github.com/elastic/beats/v7/filebeat/input/filestream/internal/task.(*Group).Go in goroutine 68
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/task/group.go:99 +0xd0

=== FAIL: filebeat/input/filestream/internal/input-logfile TestDefaultHarvesterGroup/assert_a_harvester_for_same_source_cannot_be_started (unknown)
panic: test timed out after 10m0s
	running tests:
		TestDefaultHarvesterGroup (10m0s)
		TestDefaultHarvesterGroup/assert_a_harvester_for_same_source_cannot_be_started (10m0s)

goroutine 57 [running]:
testing.(*M).startAlarm.func1()
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:2484 +0x308
created by time.goFunc
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/time/sleep.go:215 +0x38

goroutine 1 [chan receive]:
testing.(*T).Run(0x14000183500, {0x10381d7d0?, 0x1400039fb38?}, 0x103b75348)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1859 +0x388
testing.runTests.func1(0x14000183500)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:2279 +0x40
testing.tRunner(0x14000183500, 0x1400039fc68)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1792 +0xe4
testing.runTests(0x14000260dc8, {0x10425a300, 0x18, 0x18}, {0x14000374280?, 0x7?, 0x1042c3480?})
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:2277 +0x3ec
testing.(*M).Run(0x140002ca960)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:2142 +0x588
main.main()
	_testmain.go:91 +0x90

goroutine 50 [chan receive]:
testing.(*T).Run(0x140003e4540, {0x10383af10?, 0x14000080f58?}, 0x1400011c060)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1859 +0x388
github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile.TestDefaultHarvesterGroup(0x140003e4540)
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/input-logfile/harvester_test.go:238 +0x19c
testing.tRunner(0x140003e4540, 0x103b75348)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 1
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1851 +0x374

goroutine 68 [sleep]:
time.Sleep(0x2dc6c0)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/runtime/time.go:338 +0x158
github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile.TestDefaultHarvesterGroup.func6(0x14000102540)
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/input-logfile/harvester_test.go:253 +0x258
testing.tRunner(0x14000102540, 0x1400011c060)
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 50
	/Users/runner/hostedtoolcache/go/1.24.4/arm64/src/testing/testing.go:1851 +0x374

goroutine 69 [chan receive]:
github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile.blockUntilCancelOnRun({0x1400011c1f0, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {{0x0, 0x0}, {0x0, ...}, ...}, ...}, ...)
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/input-logfile/harvester_test.go:443 +0x30
github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile.(*mockHarvester).Run(_, {0x1400011c1f0, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {{0x0, 0x0}, ...}, ...}, ...)
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/input-logfile/harvester_test.go:422 +0x1e4
github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile.startHarvester.func1({0x103b88748, 0x140001320a0})
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/input-logfile/harvester.go:237 +0x7c0
github.com/elastic/beats/v7/filebeat/input/filestream/internal/task.(*Group).Go.func1()
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/task/group.go:114 +0x124
created by github.com/elastic/beats/v7/filebeat/input/filestream/internal/task.(*Group).Go in goroutine 68
	/Users/runner/work/beats/beats/filebeat/input/filestream/internal/task/group.go:99 +0xd0
```

When running `stress`, the output shows some Go internals, which seems to match some of the goroutines shown in the GitHub Actions failed run.
```
=== RUN   TestDefaultHarvesterGroup
=== RUN   TestDefaultHarvesterGroup/assert_a_harvester_is_started_in_a_goroutine
=== RUN   TestDefaultHarvesterGroup/assert_a_harvester_is_only_started_if_harvester_limit_haven't_been_reached
=== RUN   TestDefaultHarvesterGroup/assert_a_harvester_can_be_stopped_and_removed_from_bookkeeper
=== RUN   TestDefaultHarvesterGroup/assert_a_harvester_for_same_source_cannot_be_started
SIGABRT: abort
PC=0x186afbd04 m=0 sigcode=0

goroutine 0 gp=0x101dfbaa0 m=0 mp=0x101dfc520 [idle]:
runtime.kevent(0x3, 0x0, 0x0, 0x16f30e110, 0x40, 0x16f30e0e0)
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/sys_darwin.go:498 +0x20 fp=0x16f30e090 sp=0x16f30e070 pc=0x100b52350
runtime.netpoll(0x989586)
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/netpoll_kqueue.go:110 +0xf8 fp=0x16f30e920 sp=0x16f30e090 pc=0x100b2c7e8
runtime.findRunnable()
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/proc.go:3605 +0x9a4 fp=0x16f30ea90 sp=0x16f30e920 pc=0x100b39424
runtime.schedule()
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/proc.go:4020 +0x94 fp=0x16f30ead0 sp=0x16f30ea90 pc=0x100b3af44
runtime.park_m(0x1400025ce00)
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/proc.go:4144 +0x2b0 fp=0x16f30eb30 sp=0x16f30ead0 pc=0x100b3b440
runtime.mcall()
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/asm_arm64.s:193 +0x54 fp=0x16f30eb40 sp=0x16f30eb30 pc=0x100b6e734

goroutine 1 gp=0x140000021c0 m=nil [chan receive]:
runtime.gopark(0x100b71bd0?, 0x1400031f9b8?, 0xde?, 0xae?, 0x102407218?)
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/proc.go:435 +0xc8 fp=0x1400031f970 sp=0x1400031f950 pc=0x100b685c8
runtime.chanrecv(0x140000b4930, 0x1400031fa77, 0x1)
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/chan.go:664 +0x
```

```
=== RUN   TestDefaultHarvesterGroup
=== RUN   TestDefaultHarvesterGroup/assert_a_harvester_is_started_in_a_goroutine
=== RUN   TestDefaultHarvesterGroup/assert_a_harvester_is_only_started_if_harvester_limit_haven't_been_reached
=== RUN   TestDefaultHarvesterGroup/assert_a_harvester_can_be_stopped_and_removed_from_bookkeeper
=== RUN   TestDefaultHarvesterGroup/assert_a_harvester_for_same_source_cannot_be_started
=== RUN   TestDefaultHarvesterGroup/assert_a_harvester_panic_is_handled
=== RUN   TestDefaultHarvesterGroup/assert_a_harvester_error_is_handled
=== RUN   TestDefaultHarvesterGroup/assert_already_locked_resource_has_to_wait
SIGABRT: abort
PC=0x186af93cc m=0 sigcode=0

goroutine 0 gp=0x1040b7aa0 m=0 mp=0x1040b8520 [idle]:
runtime.pthread_cond_wait(0x1040b8a78, 0x1040b8a38)
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/sys_darwin.go:547 +0x20 fp=0x16d052850 sp=0x16d052820 pc=0x102e0e480
runtime.semasleep(0xffffffffffffffff)
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/os_darwin.go:72 +0x84 fp=0x16d0528b0 sp=0x16d052850 pc=0x102de8d44
runtime.notesleep(0x1040b8660)
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/lock_sema.go:62 +0xb8 fp=0x16d0528f0 sp=0x16d0528b0 pc=0x102dc22a8
runtime.mPark(...)
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/proc.go:1887
runtime.stopm()
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/proc.go:2910 +0x84 fp=0x16d052920 sp=0x16d0528f0 pc=0x102df3c84
runtime.findRunnable()
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/proc.go:3647 +0x1040 fp=0x16d052a90 sp=0x16d052920 pc=0x102df5ac0
runtime.schedule()
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.darwin-arm64/src/runtime/proc.go:4020 +0x94 fp=0x16d052ad0 sp=0x16d052a90 pc=0x102df6f44
runtime.park_m(0x140001856c0)
        /Users/tiago/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.4.da
```

--- a/filebeat/input/filestream/internal/input-logfile/harvester_test.go
+++ b/filebeat/input/filestream/internal/input-logfile/harvester_test.go
@@ -24,7 +24,6 @@ import (
 	"context"
 	"fmt"
 	"path/filepath"
-	"runtime"
 	"strings"
 	"sync"
 	"sync/atomic"
@@ -43,6 +42,23 @@ import (
 	"github.com/elastic/elastic-agent-libs/logp/logptest"
 	"github.com/elastic/elastic-agent-libs/mapstr"
 )
+
+const (
+	eventuallyTimeout  = 5 * time.Second
+	eventuallyInterval = 10 * time.Millisecond
+)
+
+// requireEventually wraps require.Eventually with consistent timeout and polling interval.
+func requireEventually(t *testing.T, condition func() bool, msgAndArgs ...any) {
+	t.Helper()
+	require.Eventually(t, condition, eventuallyTimeout, eventuallyInterval, msgAndArgs...)
+}
+
+// requireEventuallyWithT wraps require.EventuallyWithT with consistent timeout and polling interval.
+func requireEventuallyWithT(t *testing.T, condition func(c *assert.CollectT), msgAndArgs ...any) {
+	t.Helper()
+	require.EventuallyWithT(t, condition, eventuallyTimeout, eventuallyInterval, msgAndArgs...)
+}
 
 func TestReaderGroup(t *testing.T) {
 	requireGroupSuccess := func(t *testing.T, ctx context.Context, cf context.CancelFunc, err error) {
@@ -101,19 +117,11 @@ func TestReaderGroup(t *testing.T) {
 func TestDefaultHarvesterGroup(t *testing.T) {
 	source := &testSource{name: "/path/to/test"}
 
-	requireSourceAddedToBookkeeper := func(t *testing.T, hg *defaultHarvesterGroup, s Source) {
-		require.True(t, hg.readers.hasID(hg.identifier.ID(s)))
-	}
-
-	requireSourceRemovedFromBookkeeper := func(t *testing.T, hg *defaultHarvesterGroup, s Source) {
-		require.False(t, hg.readers.hasID(hg.identifier.ID(s)))
-	}
-
 	t.Run("assert a harvester is started in a goroutine", func(t *testing.T) {
 		var wg sync.WaitGroup
 
 		mockHarvester := &mockHarvester{onRun: correctOnRun, wg: &wg}
-		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		hg := testDefaultHarvesterGroup(t, mockHarvester, 0)
 
 		goroutinesChecker := resources.NewGoroutinesChecker()
 		defer goroutinesChecker.WaitUntilOriginalCount()
@@ -129,7 +137,10 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 
 		require.Equal(t, 1, mockHarvester.getRunCount())
 
-		requireSourceRemovedFromBookkeeper(t, hg, source)
+		// Wait for source to be removed from bookkeeper.
+		requireEventually(t,
+			func() bool { return !hg.readers.hasID(hg.identifier.ID(source)) },
+			"source should be removed from bookkeeper")
 		// stopped source can be stopped
 		require.Nil(t, hg.StopHarvesters())
 	})
@@ -162,8 +173,7 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 			onRun: harvesterRun,
 			wg:    &wg,
 		}
-		hg := testDefaultHarvesterGroup(t, mockHarvester)
-		hg.tg = task.NewGroup(1, time.Second, logptest.NewTestingLogger(t, ""), "")
+		hg := testDefaultHarvesterGroup(t, mockHarvester, 1)
 
 		goroutinesChecker := resources.NewGoroutinesChecker()
 		defer goroutinesChecker.WaitUntilOriginalCount()
@@ -210,8 +220,13 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 
 		require.Equal(t, 2, mockHarvester.getRunCount())
 
-		requireSourceRemovedFromBookkeeper(t, hg, source1)
-		requireSourceRemovedFromBookkeeper(t, hg, source2)
+		// Wait for sources to be removed from bookkeeper.
+		requireEventually(t,
+			func() bool {
+				return !hg.readers.hasID(hg.identifier.ID(source1)) &&
+					!hg.readers.hasID(hg.identifier.ID(source2))
+			},
+			"sources should be removed from bookkeeper")
 
 		// stopped source can be stopped
 		require.Nil(t, hg.StopHarvesters())
@@ -219,44 +234,70 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 
 	t.Run("assert a harvester can be stopped and removed from bookkeeper", func(t *testing.T) {
 		mockHarvester := &mockHarvester{onRun: blockUntilCancelOnRun}
-		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		hg := testDefaultHarvesterGroup(t, mockHarvester, 0)
 
 		goroutinesChecker := resources.NewGoroutinesChecker()
+		defer goroutinesChecker.WaitUntilOriginalCount()
 
 		ctx := input.Context{Logger: logptest.NewTestingLogger(t, ""), Cancelation: t.Context()}.WithStatusReporter(mockStatusReporter{})
 		hg.Start(ctx, source)
 
-		goroutinesChecker.WaitUntilIncreased(1)
-		// wait until harvester is started
-		require.Eventually(t,
-			func() bool { return mockHarvester.getRunCount() == 1 },
-			5*time.Second,
-			10*time.Millisecond,
-			"run count must equal one")
-		requireSourceAddedToBookkeeper(t, hg, source)
+		// Wait until harvester is started and registered in bookkeeper.
+		requireEventually(t,
+			func() bool {
+				return mockHarvester.getRunCount() == 1 && hg.readers.hasID(hg.identifier.ID(source))
+			},
+			"harvester should be running and registered")
 		// after started, stop it
 		hg.Stop(source)
-		_, err := goroutinesChecker.WaitUntilOriginalCount()
-		require.NoError(t, err)
-		requireSourceRemovedFromBookkeeper(t, hg, source)
+		// Wait for source to be removed from bookkeeper.
+		requireEventually(t,
+			func() bool { return !hg.readers.hasID(hg.identifier.ID(source)) },
+			"source should be removed from bookkeeper")
+		// StopHarvesters waits for all goroutines to fully complete, including deferred logging.
+		require.NoError(t, hg.StopHarvesters())
 	})
 
-	t.Run("assert a harvester for same source cannot be started", func(t *testing.T) {
+	sameSourceTest := func(t *testing.T, limit uint64) {
 		mockHarvester := &mockHarvester{onRun: blockUntilCancelOnRun}
-		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		hg := testDefaultHarvesterGroup(t, mockHarvester, limit)
 		inputCtx := input.Context{Logger: logptest.NewTestingLogger(t, ""), Cancelation: t.Context()}.WithStatusReporter(mockStatusReporter{})
 
 		goroutinesChecker := resources.NewGoroutinesChecker()
 		defer goroutinesChecker.WaitUntilOriginalCount()
 
+		// Start first harvester and wait for it to be registered
 		hg.Start(inputCtx, source)
-		hg.Start(inputCtx, source)
+		requireEventually(t,
+			func() bool { return hg.readers.hasID(hg.identifier.ID(source)) },
+			"first harvester should be registered")
+		// To avoid any kind of race, wait for the mock harvester to be running as well since that happens _after_ the
+		// source ID is registered
+		requireEventuallyWithT(t,
+			func(c *assert.CollectT) {
+				assert.Equal(c, 1, mockHarvester.getRunCount())
+			},
+			"one mock harvester should be running")
 
-		goroutinesChecker.WaitUntilIncreased(2)
-		// error is expected as a harvester group was expected to start twice for the same source
-		for !hg.readers.hasID(hg.identifier.ID(source)) {
+		// Repeatedly try to start additional harvesters for the same source
+		// None of them should succeed because one is already running
+		done := make(chan struct{})
+		go func() {
+			defer close(done)
+			for range 20 {
+				hg.Start(inputCtx, source)
+				time.Sleep(time.Millisecond / 4)
+			}
+		}()
+
+		// Verify runCount stays at 1 while we start new harvesters
+		for { // Check at least once
+			assert.Equal(t, 1, mockHarvester.getRunCount())
+			if isClosed(done) {
+				break
+			}
+			time.Sleep(time.Millisecond / 10) // tiny sleep to not monopolize the CPU
 		}
-		time.Sleep(3 * time.Millisecond)
 
 		hg.Stop(source)
 
@@ -264,11 +305,18 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 		require.NoError(t, err)
 
 		require.Equal(t, 1, mockHarvester.getRunCount())
+	}
+	t.Run("assert a harvester for same source cannot be started", func(t *testing.T) {
+		for _, limit := range []uint64{0, 1, 100} {
+			t.Run(fmt.Sprintf("limit=%d", limit), func(t *testing.T) {
+				sameSourceTest(t, limit)
+			})
+		}
 	})
 
 	t.Run("assert a harvester panic is handled", func(t *testing.T) {
 		mockHarvester := &mockHarvester{onRun: panicOnRun}
-		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		hg := testDefaultHarvesterGroup(t, mockHarvester, 0)
 		defer func() {
 			if v := recover(); v != nil {
 				t.Errorf("did not recover from harvester panic in defaultHarvesterGroup")
@@ -276,16 +324,21 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 		}()
 
 		goroutinesChecker := resources.NewGoroutinesChecker()
+		defer goroutinesChecker.WaitUntilOriginalCount()
 
 		ctx := input.Context{Logger: logptest.NewTestingLogger(t, ""), Cancelation: t.Context()}.WithStatusReporter(mockStatusReporter{})
 		hg.Start(ctx, source)
 
-		// wait until harvester is stopped
-		goroutinesChecker.WaitUntilOriginalCount()
+		// First wait for the harvester to run (runCount increases when Run is called,
+		// before the panic occurs).
+		requireEventually(t,
+			func() bool { return mockHarvester.getRunCount() == 1 },
+			"harvester should have run once")
 
-		// make sure harvester had run once
-		require.Equal(t, 1, mockHarvester.getRunCount())
-		requireSourceRemovedFromBookkeeper(t, hg, source)
+		// Then wait for source to be removed from bookkeeper after panic recovery.
+		requireEventually(t,
+			func() bool { return !hg.readers.hasID(hg.identifier.ID(source)) },
+			"source should be removed from bookkeeper after panic")
 
 		require.Nil(t, hg.StopHarvesters())
 	})
@@ -293,8 +346,7 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 	t.Run("assert a harvester error is handled", func(t *testing.T) {
 		testLog := &testLogger{}
 		mockHarvester := &mockHarvester{onRun: errorOnRun}
-		hg := testDefaultHarvesterGroup(t, mockHarvester)
-		hg.tg = task.NewGroup(0, 100*time.Millisecond, testLog, "")
+		hg := testDefaultHarvesterGroupWithLogger(t, mockHarvester, 0, testLog)
 
 		goroutinesChecker := resources.NewGoroutinesChecker()
 		defer goroutinesChecker.WaitUntilOriginalCount()
@@ -302,9 +354,10 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 		ctx := input.Context{Logger: logptest.NewTestingLogger(t, ""), Cancelation: t.Context()}.WithStatusReporter(mockStatusReporter{})
 		hg.Start(ctx, source)
 
-		goroutinesChecker.WaitUntilOriginalCount()
-
-		requireSourceRemovedFromBookkeeper(t, hg, source)
+		// Wait for the harvester to be removed from the bookkeeper after the error.
+		requireEventually(t,
+			func() bool { return !hg.readers.hasID(hg.identifier.ID(source)) },
+			"source should be removed from bookkeeper after error")
 
 		err := hg.StopHarvesters()
 		assert.NoError(t, err)
@@ -315,7 +368,7 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 	t.Run("assert already locked resource has to wait", func(t *testing.T) {
 		var wg sync.WaitGroup
 		mockHarvester := &mockHarvester{onRun: correctOnRun, wg: &wg}
-		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		hg := testDefaultHarvesterGroup(t, mockHarvester, 0)
 		inputCtx := input.Context{Logger: logptest.NewTestingLogger(t, ""), Cancelation: t.Context()}.WithStatusReporter(mockStatusReporter{})
 
 		r, err := lock(inputCtx, hg.store, hg.identifier.ID(source))
@@ -328,15 +381,11 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 		wg.Add(1)
 		hg.Start(inputCtx, source)
 
-		goroutinesChecker.WaitUntilIncreased(1)
-		ok := false
-		for !ok {
-			// wait until harvester is added to the bookeeper
-			ok = hg.readers.hasID(hg.identifier.ID(source))
-			if ok {
-				releaseResource(r)
-			}
-		}
+		// Wait until harvester is added to the bookkeeper.
+		requireEventually(t,
+			func() bool { return hg.readers.hasID(hg.identifier.ID(source)) },
+			"harvester should be registered in readers table")
+		releaseResource(r)
 
 		// wait until harvester.Run is done
 		wg.Wait()
@@ -349,8 +398,7 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 	t.Run("assert already locked resource has no problem when harvestergroup is cancelled", func(t *testing.T) {
 		testLog := &testLogger{}
 		mockHarvester := &mockHarvester{onRun: correctOnRun}
-		hg := testDefaultHarvesterGroup(t, mockHarvester)
-		hg.tg = task.NewGroup(0, 50*time.Millisecond, testLog, "")
+		hg := testDefaultHarvesterGroupWithLogger(t, mockHarvester, 0, testLog)
 		inputCtx := input.Context{Logger: logptest.NewTestingLogger(t, ""), Cancelation: t.Context()}.WithStatusReporter(mockStatusReporter{})
 
 		goroutinesChecker := resources.NewGoroutinesChecker()
@@ -364,7 +412,11 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 
 		hg.Start(inputCtx, source)
 
-		goroutinesChecker.WaitUntilIncreased(1)
+		// Wait for the harvester to be registered in the readers table.
+		requireEventually(t,
+			func() bool { return hg.readers.hasID(hg.identifier.ID(source)) },
+			"harvester should be registered in readers table")
+
 		assert.NoError(t, hg.StopHarvesters())
 
 		assert.Equal(t, 0, mockHarvester.getRunCount())
@@ -373,7 +425,7 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 	t.Run("assert harvester can be restarted", func(t *testing.T) {
 		var wg sync.WaitGroup
 		mockHarvester := &mockHarvester{onRun: blockUntilCancelOnRun, wg: &wg}
-		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		hg := testDefaultHarvesterGroup(t, mockHarvester, 0)
 		inputCtx := input.Context{Logger: logptest.NewTestingLogger(t, ""), Cancelation: t.Context()}.WithStatusReporter(mockStatusReporter{})
 
 		goroutinesChecker := resources.NewGoroutinesChecker()
@@ -395,60 +447,6 @@ func TestDefaultHarvesterGroup(t *testing.T) {
 		wg.Wait()
 
 		require.Equal(t, 2, mockHarvester.getRunCount())
-	})
-
-	t.Run("assert repeated start for same source does not leak goroutines with harvester_limit", func(t *testing.T) {
-		mockHarvester := &mockHarvester{onRun: blockUntilCancelOnRun}
-		hg := testDefaultHarvesterGroup(t, mockHarvester)
-		// Set harvester_limit to 1 - this creates the semaphore that caused the leak
-		hg.tg = task.NewGroup(1, time.Second, logptest.NewTestingLogger(t, ""), "")
-		inputCtx := input.Context{Logger: logptest.NewTestingLogger(t, ""), Cancelation: t.Context()}.WithStatusReporter(mockStatusReporter{})
-
-		goroutinesChecker := resources.NewGoroutinesChecker()
-
-		// Start the first harvester - this should spawn exactly 1 goroutine
-		hg.Start(inputCtx, source)
-
-		// Wait for the harvester to be running
-		require.EventuallyWithT(t,
-			func(c *assert.CollectT) {
-				assert.Equal(c, 1, mockHarvester.getRunCount())
-			},
-			5*time.Second,
-			10*time.Millisecond,
-			"harvester should be running")
-
-		// Record goroutine count after first harvester starts
-		goroutinesChecker.WaitUntilIncreased(1)
-		afterFirstStart := runtime.NumGoroutine()
-
-		// Simulate repeated file events by calling Start multiple times
-		const repeatCount = 10
-		for range repeatCount {
-			hg.Start(inputCtx, source)
-		}
-
-		// Give time for any leaked goroutines to be spawned
-		// Check that goroutine count hasn't grown significantly
-		// With the fix: should still be ~afterFirstStart (no new goroutines)
-		// Without the fix: would be afterFirstStart + repeatCount (goroutines waiting on semaphore)
-		assert.Never(t, func() bool {
-			currentGoroutines := runtime.NumGoroutine()
-			goroutineGrowth := currentGoroutines - afterFirstStart
-			return goroutineGrowth > 2
-		}, 50*time.Millisecond, 10*time.Millisecond, "goroutine count should not grow significantly after repeated Start calls")
-
-		// Cleanup
-		hg.Stop(source)
-		require.NoError(t, hg.StopHarvesters())
-
-		// Verify only 1 harvester actually ran
-		require.Equal(t, 1, mockHarvester.getRunCount(),
-			"only one harvester should have run despite multiple Start calls")
-
-		// Ensure all goroutines are cleaned up
-		_, err := goroutinesChecker.WaitUntilOriginalCount()
-		require.NoError(t, err, "all goroutines should be cleaned up")
 	})
 }
 
@@ -479,7 +477,7 @@ func TestCursorAllEventsPublished(t *testing.T) {
 
 	var cursor Cursor
 	mockHarvester := &mockHarvester{onRun: runFn, wg: &wg}
-	hg := testDefaultHarvesterGroup(t, mockHarvester)
+	hg := testDefaultHarvesterGroup(t, mockHarvester, 0)
 	hg.pipeline = &MockPipeline{
 		// Define the callback that will be called before each event is
 		// published/acknowledged, when this callback is called, the
@@ -581,14 +579,18 @@ func TestCursorAllEventsPublished(t *testing.T) {
 		"once the harvester is done, the resource must be unlocked, 'pending' must be 0")
 }
 
-func testDefaultHarvesterGroup(t *testing.T, mockHarvester Harvester) *defaultHarvesterGroup {
+func testDefaultHarvesterGroup(t *testing.T, mockHarvester Harvester, limit uint64) *defaultHarvesterGroup {
+	return testDefaultHarvesterGroupWithLogger(t, mockHarvester, limit, logptest.NewTestingLogger(t, ""))
+}
+
+func testDefaultHarvesterGroupWithLogger(t *testing.T, mockHarvester Harvester, limit uint64, logger task.Logger) *defaultHarvesterGroup {
 	return &defaultHarvesterGroup{
 		readers:    newReaderGroup(),
 		pipeline:   &MockPipeline{},
 		harvester:  mockHarvester,
 		store:      testOpenStore(t, "test", nil),
 		identifier: &SourceIdentifier{"filestream::.global::"},
-		tg:         task.NewGroup(0, time.Second, logptest.NewTestingLogger(t, ""), ""),
+		tg:         task.NewGroup(limit, 5*time.Second, logger, ""),
 	}
 }
 
@@ -745,4 +747,13 @@ type mockStatusReporter struct{}
 
 // UpdateStatus is a no-op
 func (m mockStatusReporter) UpdateStatus(status status.Status, msg string) {
+}
+
+func isClosed(done chan struct{}) bool {
+	select {
+	case <-done:
+		return true
+	default:
+	}
+	return false
 }

--- a/libbeat/tests/resources/goroutines.go
+++ b/libbeat/tests/resources/goroutines.go
@@ -102,11 +102,3 @@ func (c *GoroutinesChecker) WaitUntilOriginalCount() (int, error) {
 	}
 	return after, ErrTimeout
 }
-
-// WaitUntilIncreased waits till the number of goroutines is n plus the number
-// before creating the checker.
-func (c *GoroutinesChecker) WaitUntilIncreased(n int) {
-	for runtime.NumGoroutine() < c.before+n {
-		time.Sleep(10 * time.Millisecond)
-	}
-}

